### ISSUE TITLE
aria2 crashes

### ISSUE THREAD

**Issue by Denisov39** [aria2/aria2#725]
aria 1.25.0
** Configuration **
Enabled Features: Async DNS, BitTorrent, Firefox3 Cookie, GZip, HTTPS, Message Digest, Metalink, XML-RPC, SFTP
Hash Algorithms: sha-1, sha-224, sha-256, sha-384, sha-512, md5, adler32
Libraries: zlib/1.2.8 libxml2/2.9.1 sqlite3/3.8.7.1 OpenSSL/1.0.1t c-ares/1.10.0 libssh2/1.4.3
Compiler: gcc 4.9.2
  built by   x86_64-unknown-linux-gnu
  on         Aug 18 2016 16:51:31
System: Linux 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u4 (2016-02-29) x86_6
```
#0  0x00007f0fa82c9067 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f0fa82ca448 in __GI_abort () at abort.c:89
#2  0x00007f0fa82c2266 in __assert_fail_base (fmt=0x7f0faa32f4f3 "%s%s%s:%u: %s%sПроверочное утверждение «%s» не выполнено.\n%n", 
    assertion=assertion@entry=0x76aec0 "uniqPeers_.size() == unusedPeers_.size() + usedPeers_.size()", file=file@entry=0x76aea3 "DefaultPeerStorage.cc", line=line@entry=70, 
    function=function@entry=0x76b160 <aria2::DefaultPeerStorage::~DefaultPeerStorage()::__PRETTY_FUNCTION__> "virtual aria2::DefaultPeerStorage::~DefaultPeerStorage()") at assert.c:92
#3  0x00007f0fa82c2312 in __GI___assert_fail (assertion=0x76aec0 "uniqPeers_.size() == unusedPeers_.size() + usedPeers_.size()", file=0x76aea3 "DefaultPeerStorage.cc", line=70, 
    function=0x76b160 <aria2::DefaultPeerStorage::~DefaultPeerStorage()::__PRETTY_FUNCTION__> "virtual aria2::DefaultPeerStorage::~DefaultPeerStorage()") at assert.c:101
#4  0x000000000052e9fa in aria2::DefaultPeerStorage::~DefaultPeerStorage (this=0x2e64b20, __in_chrg=<optimized out>) at DefaultPeerStorage.cc:70
#5  0x000000000048bb7f in __gnu_cxx::new_allocator<aria2::DefaultPeerStorage>::destroy<aria2::DefaultPeerStorage> (this=0x2e64b20, __p=0x2e64b20) at /usr/include/c++/4.9/ext/new_allocator.h:124
#6  0x000000000048b885 in std::allocator_traits<std::allocator<aria2::DefaultPeerStorage> >::_S_destroy<aria2::DefaultPeerStorage> (__a=..., __p=0x2e64b20) at /usr/include/c++/4.9/bits/alloc_traits.h:282
#7  0x000000000048b4a5 in std::allocator_traits<std::allocator<aria2::DefaultPeerStorage> >::destroy<aria2::DefaultPeerStorage> (__a=..., __p=0x2e64b20) at /usr/include/c++/4.9/bits/alloc_traits.h:411
#8  0x000000000048ac5f in std::_Sp_counted_ptr_inplace<aria2::DefaultPeerStorage, std::allocator<aria2::DefaultPeerStorage>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x2e64b10)
    at /usr/include/c++/4.9/bits/shared_ptr_base.h:524
#9  0x000000000040e1b6 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x2e64b10) at /usr/include/c++/4.9/bits/shared_ptr_base.h:149
#10 0x000000000040df71 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x32d9bb0, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/shared_ptr_base.h:666
#11 0x000000000047845a in std::__shared_ptr<aria2::PeerStorage, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x32d9ba8, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/shared_ptr_base.h:914
#12 0x0000000000478474 in std::shared_ptr<aria2::PeerStorage>::~shared_ptr (this=0x32d9ba8, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/shared_ptr.h:93
#13 0x00000000005a6878 in aria2::DefaultBtProgressInfoFile::~DefaultBtProgressInfoFile (this=0x32d9b80, __in_chrg=<optimized out>) at DefaultBtProgressInfoFile.cc:87
#14 0x000000000048bbff in __gnu_cxx::new_allocator<aria2::DefaultBtProgressInfoFile>::destroy<aria2::DefaultBtProgressInfoFile> (this=0x32d9b80, __p=0x32d9b80) at /usr/include/c++/4.9/ext/new_allocator.h:124
#15 0x000000000048b935 in std::allocator_traits<std::allocator<aria2::DefaultBtProgressInfoFile> >::_S_destroy<aria2::DefaultBtProgressInfoFile> (__a=..., __p=0x32d9b80)
    at /usr/include/c++/4.9/bits/alloc_traits.h:282
#16 0x000000000048b56d in std::allocator_traits<std::allocator<aria2::DefaultBtProgressInfoFile> >::destroy<aria2::DefaultBtProgressInfoFile> (__a=..., __p=0x32d9b80)
    at /usr/include/c++/4.9/bits/alloc_traits.h:411
#17 0x000000000048af1b in std::_Sp_counted_ptr_inplace<aria2::DefaultBtProgressInfoFile, std::allocator<aria2::DefaultBtProgressInfoFile>, (__gnu_cxx::_Lock_policy)2>::_M_dispose (this=0x32d9b70)
    at /usr/include/c++/4.9/bits/shared_ptr_base.h:524
#18 0x000000000040e1b6 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x32d9b70) at /usr/include/c++/4.9/bits/shared_ptr_base.h:149
#19 0x000000000040df71 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count (this=0x7ffcb5601d68, __in_chrg=<optimized out>) at /usr/include/c++/4.9/bits/shared_ptr_base.h:666
#20 0x0000000000478026 in std::__shared_ptr<aria2::BtProgressInfoFile, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr (this=0x7ffcb5601d60, __in_chrg=<optimized out>)
    at /usr/include/c++/4.9/bits/shared_ptr_base.h:914
#21 0x000000000047f90d in std::__shared_ptr<aria2::BtProgressInfoFile, (__gnu_cxx::_Lock_policy)2>::operator=<aria2::NullProgressInfoFile>(std::__shared_ptr<aria2::NullProgressInfoFile, (__gnu_cxx::_Lock_policy)2>&&) (this=0x43415e8, __r=<unknown type in /home/denisov/aria2c, CU 0x24fbd4, DIE 0x2e32bb>) at /usr/include/c++/4.9/bits/shared_ptr_base.h:1001
#22 0x000000000047bf86 in std::shared_ptr<aria2::BtProgressInfoFile>::operator=<aria2::NullProgressInfoFile>(std::shared_ptr<aria2::NullProgressInfoFile>&&) (this=0x43415e8, 
    __r=<unknown type in /home/denisov/aria2c, CU 0x24fbd4, DIE 0x2dc6b5>) at /usr/include/c++/4.9/bits/shared_ptr.h:299
#23 0x0000000000475bbd in aria2::RequestGroup::releaseRuntimeResource (this=0x4341580, e=0x21ae4e0) at RequestGroup.cc:1023
#24 0x000000000048dadc in aria2::(anonymous namespace)::ProcessStoppedRequestGroup::operator() (this=0x7ffcb5601fc0, group=std::shared_ptr (count 2, weak 0) 0x4341580) at RequestGroupMan.cc:460
#25 0x000000000049121d in aria2::IndexedList<unsigned long, std::shared_ptr<aria2::RequestGroup> >::remove_if<aria2::(anonymous namespace)::ProcessStoppedRequestGroup> (this=0x1ff6e90, pred=...)
    at IndexedList.h:471
#26 0x000000000048ddda in aria2::RequestGroupMan::removeStoppedGroup (this=0x1ff6e90, e=0x21ae4e0) at RequestGroupMan.cc:478
#27 0x000000000048e1ab in aria2::RequestGroupMan::fillRequestGroupFromReserver (this=0x1ff6e90, e=0x21ae4e0) at RequestGroupMan.cc:517
#28 0x00000000005d58ca in aria2::FillRequestGroupCommand::execute (this=0x24fe100) at FillRequestGroupCommand.cc:70
#29 0x00000000005badf5 in aria2::(anonymous namespace)::executeCommand (commands=std::deque with 7 elements = {...}, statusFilter=aria2::Command::STATUS_ALL) at DownloadEngine.cc:139
#30 0x00000000005bafcf in aria2::DownloadEngine::run (this=0x21ae4e0, oneshot=false) at DownloadEngine.cc:177
#31 0x0000000000452ab5 in aria2::MultiUrlRequestInfo::execute (this=0x1b369c0) at MultiUrlRequestInfo.cc:361
#32 0x000000000040d458 in aria2::main (argc=2, argv=0x7ffcb56025c8) at main.cc:78
#33 0x000000000040d4e2 in main (argc=2, argv=0x7ffcb56025c8) at main.cc:91
```

**Comment by tatsuhiro-t** [comment]
Thank you for reporting this issue.
Fix committed via  65b7205dfdc182a0d6a694547b2d5710dfd7466d

**Comment by tatsuhiro-t** [comment]
Reverted the above commit, and fixed it in another way to leave assertion there.  See 7a31f9bf29c6b89aed2a216d0159d6f43369e8a2
